crates/file-serve/src/lib.rs

```diff
@@ -100,17 +105,54 @@ impl Server {
         self.addr.as_str()
     }
 
+    /// Whether the server was running at the instant the call happened
+    pub fn is_running(&self) -> bool {
+        self.running.load(Ordering::Acquire)
+    }
+
     /// Start the webserver
-    pub fn serve(&self) -> Result<std::convert::Infallible, Error> {
-        // attempts to create a server
-        let server = tiny_http::Server::http(self.addr()).map_err(Error::new)?;
+    pub fn serve(&self) -> Result<(), Error> {
+        if self
+            .running
+            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
+            .is_err()
+        {
+            return Err(Error::new("the server is running"));
+        }
 
-        for request in server.incoming_requests() {
-            if let Err(e) = static_file_handler(self.source(), request) {
-                log::error!("{}", e);
+        // attempts to create a server
+        let server = match tiny_http::Server::http(self.addr()) {
+            Ok(server) => server,
+            Err(e) => {
+                self.running.store(false, Ordering::Release);
+                return Err(Error::new(e));
```

Comment:
nit: Since I'm already making another comment, I worry about keeping `self.running` valid through all of the error handling.

One way of dealing with that is to have a `serve_` function with the core logic and `serve` calls it and handles the `store(false)`.  This will ensure it will always be reverted, no matter the code path.  I guess an RAII object might also work